src/iceberg/snapshot.h

```diff
@@ -0,0 +1,224 @@
+  /// \brief The original id of a cherry-picked snapshot
+  constexpr static std::string_view kSourceSnapshotID = "source-snapshot-id";
+  /// \brief Name of the engine that created the snapshot
+  constexpr static std::string_view kEngineName = "engine-name";
+  /// \brief Version of the engine that created the snapshot
+  constexpr static std::string_view kEngineVersion = "engine-version";
+};
+
+/// \brief Data operation that produce snapshots.
+///
+/// A snapshot can return the operation that created the snapshot to help other components
+/// ignore snapshots that are not needed for some tasks. For example, snapshot expiration
+/// does not need to clean up deleted files for appends, which have no deleted files.
+struct ICEBERG_EXPORT DataOperation {
+  /// \brief Only data files were added and no files were removed.
+  static constexpr std::string_view kAppend = "append";
+  /// \brief Data and delete files were added and removed without changing table data;
+  /// i.e. compaction, change the data file format, or relocating data files.
+  static constexpr std::string_view kReplace = "replace";
+  /// \brief Data and delete files were added and removed in a logical overwrite
+  /// operation.
+  static constexpr std::string_view kOverwrite = "overwrite";
+  /// \brief Data files were removed and their contents logically deleted and/or delete
+  /// files were added to delete rows.
+  static constexpr std::string_view kDelete = "delete";
+};
+
+/// \brief A snapshot of the data in a table at a point in time.
+///
+/// A snapshot consist of one or more file manifests, and the complete table contents is
+/// the union of all the data files in those manifests.
+///
+/// Snapshots are created by table operations.
+struct ICEBERG_EXPORT Snapshot {
+  using manifest_list_t = std::string;
+  using manifests_t = std::vector<std::string>;
+
+  /// A unqiue long ID.
+  int64_t snapshot_id;
+  /// The snapshot ID of the snapshot's parent. Omitted for any snapshot with no parent.
+  std::optional<int64_t> parent_snapshot_id;
+  /// A monotonically increasing long that tracks the order of changes to a table.
+  int64_t sequence_number;
+  /// A timestamp when the snapshot was created, used for garbage collection and table
+  /// inspection.
+  int64_t timestamp_ms;
+  /// The location of a manifest list for this snapshot that tracks manifest files with
+  /// additional metadata.
+  std::variant<manifest_list_t, manifests_t> manifest_list;
+  /// A string map that summaries the snapshot changes, including operation.
+  std::unordered_map<std::string, std::string> summary;
+  /// ID of the table's current schema when the snapshot was created.
+  std::optional<int32_t> schema_id;
+
+  /// \brief Return the name of the DataOperations data operation that produced this
+  /// snapshot.
+  ///
+  /// \return the operation that produced this snapshot, or nullopt if the operation is
+  /// unknown.
+  std::optional<std::string> operation() const;
```

Comment:
nit: maybe string_view? Assuming this is a reference into `summary`

Reply:
done.